src/devtools/client/webconsole/components/Input/JSTerm.tsx

```diff
@@ -0,0 +1,185 @@
+  if (matchCount === 1 && insertAutocompleteMatch(expression, matches[0]) === expression) {
+    return false;
+  }
+
+  return !hideAutocomplete && matchCount;
+}
+
+function useHistory(setValue: Dispatch<SetStateAction<string>>) {
+  const commandHistory = useSelector(getCommandHistory);
+  const [historyIndex, setHistoryIndex] = useState<number>(0);
+
+  const moveHistoryCursor = (difference: number) => {
+    if (commandHistory.length > 0) {
+      const newIndex = clamp(historyIndex + difference, 0, commandHistory.length);
+
+      setValue(["", ...commandHistory][newIndex]);
+      setHistoryIndex(newIndex);
+    }
+  };
+
+  return { moveHistoryCursor, setHistoryIndex };
+}
+
+function useAutocomplete(expression: string) {
+  const [hideAutocomplete, setHideAutocomplete] = useState<boolean>(true);
+  const [autocompleteIndex, setAutocompleteIndex] = useState<number>(0);
+  const showAutocomplete = useShowAutocomplete(expression, hideAutocomplete);
+  const matches = useGetMatches(expression);
+
+  const moveAutocompleteCursor = (difference: number) => {
+    const matchesCount = matches.length;
+
+    const newIndex = (matchesCount + autocompleteIndex - difference) % matchesCount;
+    setAutocompleteIndex(newIndex);
+  };
+  const applySelectedMatch = () => {
+    const match = matches[autocompleteIndex];
+    return insertAutocompleteMatch(expression, match);
+  };
+
+  return {
+    applySelectedMatch,
+    autocompleteIndex,
+    matches,
+    moveAutocompleteCursor,
+    setAutocompleteIndex,
+    setHideAutocomplete,
+    showAutocomplete,
+  };
+}
+
+export default function JSTerm() {
+  const dispatch = useDispatch();
+  const recordingId = useGetRecordingId();
+  const { recording } = useGetRecording(recordingId);
+
+  const [value, setValue] = useState<string>("");
+  const inputNode = useRef<HTMLDivElement | null>(null);
+
+  const { moveHistoryCursor, setHistoryIndex } = useHistory(setValue);
+  const {
+    applySelectedMatch,
+    autocompleteIndex,
+    matches,
+    moveAutocompleteCursor,
+    setAutocompleteIndex,
+    setHideAutocomplete,
+    showAutocomplete,
+  } = useAutocomplete(value);
+
+  const onRegularKeyPress = (e: KeyboardEvent) => {
+    if (e.key === "Enter") {
+      execute();
+    } else if (e.key === "ArrowUp") {
+      moveHistoryCursor(1);
+    } else if (e.key === "ArrowDown") {
+      moveHistoryCursor(-1);
+    }
+  };
+  const onAutocompleteKeyPress = (e: KeyboardEvent) => {
+    e.preventDefault();
+
+    if (e.key === "Enter" || e.key === "Tab") {
+      autocomplete();
+    } else if (e.key === "ArrowDown") {
+      moveAutocompleteCursor(-1);
+    } else if (e.key === "ArrowUp") {
+      moveAutocompleteCursor(1);
+    }
+
+    if (["Enter", "Tab", "Escape", "ArrowRight", "ArrowLeft"].includes(e.key)) {
+      setHideAutocomplete(true);
+    } else if (!["ArrowDown", "ArrowUp"].includes(e.key)) {
+      setHideAutocomplete(false);
+      setAutocompleteIndex(0);
+    }
+  };
```

Comment:
No magic strings please, let's move all of these to consts that live at the top of the file and can be easily updated if needed.